### PR TITLE
Add quantization techniques and links to byte vector

### PR DESCRIPTION
**Pull request by kolchfa-aws**
Add quantization techniquest and links to byte vector

### Checklist
- [x] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: _field-types/supported-field-types/knn-vector.md [hunk 1]
```diff
@@ -163,4 +163,40 @@ GET test-index/_search
   }
 }
 ```
-{% include copy-curl.html %}
\ No newline at end of file
+{% include copy-curl.html %}
+
+### Quantization techniques
```
Comment (naveentatikonda) [inline _field-types/supported-field-types/knn-vector.md:168]: Can we also include the pseudo code used for ConsineSimilarity SpaceType and the reason behind using a different technique for different type of dataset(angular) which emphasizes the importance of a quantization technique based on type of data?

**Review by naveentatikonda — CHANGES_REQUESTED** [review]

File: _field-types/supported-field-types/knn-vector.md (hunk 1 above)
Comment (kolchfa-aws) [reply _field-types/supported-field-types/knn-vector.md:171]: @naveentatikonda Done. Please review when you get a chance.

File: _field-types/supported-field-types/knn-vector.md [hunk 2]
```diff
@@ -163,4 +166,102 @@ GET test-index/_search
...
+
+#### Scalar quantization for the L2 space type
+
+For Euclidean datasets, we recommend using a scalar quantization technique with L2 space type because Euclidean distance is shift invariant. If you shift both $$x$$ and $$y$$ by the same $$z$$ then the distance remains the same, which means $$\lVert x-y\rVert =\lVert (x-z)-(y-z)\rVert$$.
+
+The following example pseudocode illustrates scalar quantization for the L2 space type:
```
Comment (naveentatikonda) [inline _field-types/supported-field-types/knn-vector.md:179]: We shouldn't explain it this way. Can you rephrase it to something like this:

The following example pseudocode illustrates the scalar quantization technique used for the benchmarking tests on Euclidean Datasets (uses L2 space type). Euclidean distance is shift invariant, if you shift both $$x$$ and $$y$$ by the same $$z$$ then the distance remains the same, which means $$\lVert x-y\rVert =\lVert (x-z)-(y-z)\rVert$$.

File: _field-types/supported-field-types/knn-vector.md [hunk 3]
```diff
@@ -163,4 +166,102 @@ GET test-index/_search
...
+# Bucket into 256 values
+queryset = np.floor(queryset * (B - 1)) - int(B / 2)
+```
+{% include copy.html %}
+
+#### Scalar quantization for the cosine similarity space type
+
+For angular datasets, we recommend using a scalar quantization technique with cosine similarity because cosine distance is not shift invariant ($$cos(x, y) \neq cos(x-z, y-z)$$). 
+
+The following example pseudocode illustrates scalar quantization for the cosine similarity space type:
```
Comment (naveentatikonda) [inline _field-types/supported-field-types/knn-vector.md:216]: Similarly, we need to rephrase this:

The following example pseudocode illustrates the scalar quantization technique used for the benchmarking tests on Angular Datasets (uses cosine similarity space type). Cosine similarity is not shift invariant ($$cos(x, y) \neq cos(x-z, y-z)$$).

File: _field-types/supported-field-types/knn-vector.md [hunk 4]
```diff
@@ -163,4 +166,102 @@ GET test-index/_search
...
+frac_part = val - int_part
+
+if 0.5 < frac_part:
+ bval = int_part + 1
+else:
+ bval = int_part
+
+return Byte(bval)
+
+// For Negative Numbers
```
Comment (naveentatikonda) [inline _field-types/supported-field-types/knn-vector.md:243]: From `For Negative Numbers`, can we break it into separate code block to make it easy to understand.

Comment (naveentatikonda) [inline _field-types/supported-field-types/knn-vector.md:243]: nit: 
```suggestion
# For Negative Numbers
```

File: _field-types/supported-field-types/knn-vector.md [hunk 5]
```diff
@@ -163,4 +166,102 @@ GET test-index/_search
...
 }
 ```
-{% include copy-curl.html %}
\ No newline at end of file
+{% include copy-curl.html %}
+
+### Quantization techniques
+
+If your vectors are of type `float`, you need to first convert them to `byte` before ingesting the documents. This conversion is accomplished by _quantizing the dataset_---reducing the precision of its vectors. There are many quantization techniques, such as scalar quantization or product quantization (PQ), which is used in the Faiss engine. The choice of quantization technique depends on the type of data you're using and can affect the accuracy of recall values. The following sections describe the scalar quantization algorithms that were used to quantize the [k-NN benchmarking test](https://github.com/opensearch-project/k-NN/tree/main/benchmarks/perf-tool) data for the [L2](#scalar-quantization-for-the-l2-space-type) and [cosine similarity](#scalar-quantization-for-the-cosine-similarity-space-type) space types.
+
```
Comment (naveentatikonda) [inline _field-types/supported-field-types/knn-vector.md:174]: Can we add a note or statement saying that the below quantization techniques are for reference

**Review by naveentatikonda — CHANGES_REQUESTED** [review]

File: _field-types/supported-field-types/knn-vector.md (hunk 1 above)
Comment (naveentatikonda) [reply _field-types/supported-field-types/knn-vector.md:171]: @kolchfa-aws Added few comments. Pls take a look

**Comment by naveentatikonda** [comment]
PR Looks Good. @kolchfa-aws Thanks for making the changes

**Review by naveentatikonda — APPROVED** [review]

File: _field-types/supported-field-types/knn-vector.md [hunk 6]
```diff
@@ -163,4 +166,105 @@ GET test-index/_search
   }
 }
 ```
-{% include copy-curl.html %}
\ No newline at end of file
+{% include copy-curl.html %}
+
+### Quantization techniques
+
+If your vectors are of type `float`, you need to first convert them to `byte` before ingesting the documents. This conversion is accomplished by _quantizing the dataset_---reducing the precision of its vectors. There are many quantization techniques, such as scalar quantization or product quantization (PQ), which is used in the Faiss engine. The choice of quantization technique depends on the type of data you're using and can affect the accuracy of recall values. The following sections describe the scalar quantization algorithms that were used to quantize the [k-NN benchmarking test](https://github.com/opensearch-project/k-NN/tree/main/benchmarks/perf-tool) data for the [L2](#scalar-quantization-for-the-l2-space-type) and [cosine similarity](#scalar-quantization-for-the-cosine-similarity-space-type) space types. The provided pseudocode is for illustration only.
```
Comment (Naarcha-AWS) [inline _field-types/supported-field-types/knn-vector.md:173]:
```suggestion
If your vectors are of the type `float`, you need to first convert them to the `byte` type before ingesting the documents. This conversion is accomplished by _quantizing the dataset_---reducing the precision of its vectors. There are many quantization techniques, such as scalar quantization or product quantization (PQ), which is used in the Faiss engine. The choice of quantization technique depends on the type of data you're using and can affect the accuracy of recall values. The following sections describe the scalar quantization algorithms that were used to quantize the [k-NN benchmarking test](https://github.com/opensearch-project/k-NN/tree/main/benchmarks/perf-tool) data for the [L2](#scalar-quantization-for-the-l2-space-type) and [cosine similarity](#scalar-quantization-for-the-cosine-similarity-space-type) space types. The provided pseudocode is for illustration only.
```

File: _field-types/supported-field-types/knn-vector.md [hunk 7]
```diff
@@ -163,4 +166,105 @@ GET test-index/_search
...
+
+#### Scalar quantization for the L2 space type
+
+The following example pseudocode illustrates the scalar quantization technique used for the benchmarking tests on Euclidean datasets with the L2 space type. Euclidean distance is shift invariant. If you shift both $$x$$ and $$y$$ by the same $$z$$ then the distance remains the same ($$\lVert x-y\rVert =\lVert (x-z)-(y-z)\rVert$$).
```
Comment (Naarcha-AWS) [inline _field-types/supported-field-types/knn-vector.md:177]: A shift invariant? Should we define shift invariant of do we do so earlier?

**Review by Naarcha-AWS — APPROVED** [review]
A few small comments, otherwise LGTM

File: _field-types/supported-field-types/knn-vector.md (hunk 7 above)
Comment (kolchfa-aws) [reply _field-types/supported-field-types/knn-vector.md:177]: It's defined in the following sentence.

File: _field-types/supported-field-types/knn-vector.md [hunk 8]
```diff
@@ -4,11 +4,12 @@ title: k-NN vector
 nav_order: 58
 has_children: false
 parent: Supported field types
+has_math: true
 ---
 
 # k-NN vector 
 
-The k-NN plugin introduces a custom data type, the `knn_vector`, that allows users to ingest their k-NN vectors
+The [k-NN plugin]({{site.url}}{{site.baseurl}}/search-plugins/knn/index/) introduces a custom data type, the `knn_vector`, that allows users to ingest their k-NN vectors
```
Comment (natebower) [inline _field-types/supported-field-types/knn-vector.md:12]:
```suggestion
The [k-NN plugin]({{site.url}}{{site.baseurl}}/search-plugins/knn/index/) introduces a custom data type, the `knn_vector`, that allows users to ingest their k-NN vectors.
```

File: _field-types/supported-field-types/knn-vector.md [hunk 9]
```diff
@@ -47,7 +48,7 @@ PUT test-index
 
 ## Method definitions
 
-Method definitions are used when the underlying Approximate k-NN algorithm does not require training. For example, the following `knn_vector` field specifies that *nmslib*'s implementation of *hnsw* should be used for Approximate k-NN search. During indexing, *nmslib* will build the corresponding *hnsw* segment files.
+[Method definitions]({{site.url}}{{site.baseurl}}/search-plugins/knn/knn-index#method-definitions) are used when the underlying [Approximate k-NN]({{site.url}}{{site.baseurl}}/search-plugins/knn/approximate-knn/) algorithm does not require training. For example, the following `knn_vector` field specifies that *nmslib*'s implementation of *hnsw* should be used for Approximate k-NN search. During indexing, *nmslib* will build the corresponding *hnsw* segment files.
```
Comment (natebower) [inline _field-types/supported-field-types/knn-vector.md:51]:
```suggestion
[Method definitions]({{site.url}}{{site.baseurl}}/search-plugins/knn/knn-index#method-definitions) are used when the underlying [approximate k-NN]({{site.url}}{{site.baseurl}}/search-plugins/knn/approximate-knn/) algorithm does not require training. For example, the following `knn_vector` field specifies that *nmslib*'s implementation of *hnsw* should be used for approximate k-NN search. During indexing, *nmslib* will build the corresponding *hnsw* segment files.
```

File: _field-types/supported-field-types/knn-vector.md [hunk 10]
```diff
@@ -77,7 +78,7 @@ model contains the information needed to initialize the native library segment f
 }
 ```
 
-However, if you intend to just use painless scripting or a k-NN score script, you only need to pass the dimension.
+However, if you intend to just use Painless scripting or a k-NN score script, you only need to pass the dimension.
```
Comment (natebower) [inline _field-types/supported-field-types/knn-vector.md:81]: I would either replace "just" with "only" or delete "just".

File: _field-types/supported-field-types/knn-vector.md [hunk 11]
```diff
@@ -91,6 +92,8 @@ By default, k-NN vectors are `float` vectors, where each dimension is 4 bytes. I
 Byte vectors are supported only for the `lucene` engine. They are not supported for the `nmslib` and `faiss` engines.
 {: .note}
 
+In [k-NN benchmarking tests](https://github.com/opensearch-project/k-NN/tree/main/benchmarks/perf-tool), the use of `byte` rather than `float` vectors resulted in a significant reduction in storage and memory usage while also improving indexing throughput and reducing query latency. Additionally, precision on recall was not greatly affected (note that recall can depend on various factors, such as the [quantization technique](#quantization-techniques) used and data distribution).
```
Comment (natebower) [inline _field-types/supported-field-types/knn-vector.md:95]:
```suggestion
In [k-NN benchmarking tests](https://github.com/opensearch-project/k-NN/tree/main/benchmarks/perf-tool), the use of `byte` rather than `float` vectors resulted in a significant reduction in storage and memory usage as well as improved indexing throughput and reduced query latency. Additionally, precision on recall was not greatly affected (note that recall can depend on various factors, such as the [quantization technique](#quantization-techniques) used and data distribution). 
```

Comment (natebower) [inline _field-types/supported-field-types/knn-vector.md:95]: This reads as though there should be some words before "data distribution". Something like "such as the quantization technique used and the type of data distribution."

File: _field-types/supported-field-types/knn-vector.md [hunk 12]
```diff
@@ -163,4 +166,105 @@ GET test-index/_search
   }
 }
 ```
-{% include copy-curl.html %}
\ No newline at end of file
+{% include copy-curl.html %}
+
+### Quantization techniques
+
+If your vectors are of the type `float`, you need to first convert them to the `byte` type before ingesting the documents. This conversion is accomplished by _quantizing the dataset_---reducing the precision of its vectors. There are many quantization techniques, such as scalar quantization or product quantization (PQ), which is used in the Faiss engine. The choice of quantization technique depends on the type of data you're using and can affect the accuracy of recall values. The following sections describe the scalar quantization algorithms that were used to quantize the [k-NN benchmarking test](https://github.com/opensearch-project/k-NN/tree/main/benchmarks/perf-tool) data for the [L2](#scalar-quantization-for-the-l2-space-type) and [cosine similarity](#scalar-quantization-for-the-cosine-similarity-space-type) space types. The provided pseudocode is for illustration only.
```
Comment (natebower) [inline _field-types/supported-field-types/knn-vector.md:173]:
```suggestion
If your vectors are of the type `float`, you need to first convert them to the `byte` type before ingesting the documents. This conversion is accomplished by _quantizing the dataset_---reducing the precision of its vectors. There are many quantization techniques, such as scalar quantization or product quantization (PQ), which is used in the Faiss engine. The choice of quantization technique depends on the type of data you're using and can affect the accuracy of recall values. The following sections describe the scalar quantization algorithms that were used to quantize the [k-NN benchmarking test](https://github.com/opensearch-project/k-NN/tree/main/benchmarks/perf-tool) data for the [L2](#scalar-quantization-for-the-l2-space-type) and [cosine similarity](#scalar-quantization-for-the-cosine-similarity-space-type) space types. The provided pseudocode is for illustration purposes only.
```

File: _field-types/supported-field-types/knn-vector.md [hunk 13]
```diff
@@ -163,4 +166,105 @@ GET test-index/_search
...
+
+#### Scalar quantization for the L2 space type
+
+The following example pseudocode illustrates the scalar quantization technique used for the benchmarking tests on Euclidean datasets with the L2 space type. Euclidean distance is shift invariant. If you shift both $$x$$ and $$y$$ by the same $$z$$ then the distance remains the same ($$\lVert x-y\rVert =\lVert (x-z)-(y-z)\rVert$$).
```
Comment (natebower) [inline _field-types/supported-field-types/knn-vector.md:177]:
```suggestion
The following example pseudocode illustrates the scalar quantization technique used for the benchmarking tests on Euclidean datasets with the L2 space type. Euclidean distance is shift invariant. If you shift both $$x$$ and $$y$$ by the same $$z$$, then the distance remains the same ($$\lVert x-y\rVert =\lVert (x-z)-(y-z)\rVert$$).
```

File: _search-plugins/knn/knn-index.md [hunk 14]
```diff
@@ -11,6 +11,10 @@ has_children: false
 The k-NN plugin introduces a custom data type, the `knn_vector`, that allows users to ingest their k-NN vectors
 into an OpenSearch index and perform different kinds of k-NN search. The `knn_vector` field is highly configurable and can serve many different k-NN workloads. For more information, see [k-NN vector]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-vector/).
 
+## Lucene byte vector
+
+Starting with k-NN plugin version 2.9, you can use `byte` vectors with the `lucene` engine in order to save storage space. For more information, see [Lucene byte vector]({{site.url}}{{site.baseurl}}/field-types/supported-field-types/knn-vector#lucene-byte-vector).
```
Comment (natebower) [inline _search-plugins/knn/knn-index.md:16]: Something like "reduce the amount of storage space needed" instead of "save storage space"? To "save storage space" means to set it aside for another purpose.

**Review by natebower — COMMENTED** [review]
@kolchfa-aws Please see my comments and changes and let me know if you have any questions or would like me to read any revisions. Thanks!

File: _field-types/supported-field-types/knn-vector.md (hunk 10 above)
Comment (kolchfa-aws) [inline _field-types/supported-field-types/knn-vector.md:81]:
```suggestion
However, if you intend to use Painless scripting or a k-NN score script, you only need to pass the dimension.
```

File: _field-types/supported-field-types/knn-vector.md (hunk 11 above)
Comment (kolchfa-aws) [reply _field-types/supported-field-types/knn-vector.md]: Reworded.
